Redesign Dired link menu and cleanup Makefiles

kickingvegas opened the pull request:
- Redesign link menu to support the following types of directory links:
  - Absolute Symbolic
  - Relative Symbolic
  - Hard
- Remove "Symlink…" (S) menu item from `casual-dired-tmenu`.
- Add "Link›" (l) menu item to `casual-dired-tmenu`.
- Update screenshots and documentation.
- Cleanup Makefiles to consolidate common variables, rules, and targets.

kickingvegas commented:
@purcell - Thanks for the input and TIL about `autoload`. Not having to call `require` to init Casual would be nice. That said, my biggest concern is knowing how to test `autoload` before releasing it.

Also, am I correct in understanding that folks who install Casual from the repo would still need to invoke `(require 'casual)`?

As far as development practice, I use the `main` branch as a release branch and `development` as a staging branch for the next release. If you don’t mind, please reconfigure your PR to target `development`.

purcell commented:
> That said, my biggest concern is knowing how to test autoload before releasing it.

Testing such things automatically is not something I've generally seen done, but you can generate the autoloads file yourself, in the same way that the package installation process does it, by calling `update-directory-autoloads` (or `loaddefs-generate` in Emacs 29.x). This gives you an autoload file that you can load. You shouldn't commit that file, though, because its contents can end up being specific to the emacs version you used to creat it. More on the autoload process here: https://www.gnu.org/software/emacs/manual/html_node/elisp/Autoload.html

> Also, am I correct in understanding that folks who install Casual from the repo would still need to invoke (require 'casual)?

Depends what you mean by "install". Most installation mechanisms generate the autoloads file for the user. But certainly if someone just includes the Casual repo as, say, a git submodule, then they wouldn't have an autoload file and would still need to `require` the module.

> If you don’t mind, please reconfigure your PR to target development.

Sure, I'll see if I can do this now.

purcell commented:
I can't change the target branch without recreating the PR, but you should be able to do so easily: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-base-branch-of-a-pull-request

(I generally suggest making a repo's default branch the development branch, otherwise you'll find this to be an ongoing issue.)

kickingvegas commented:
@purcell I've thought of making the repo development branch the default one; however when trying to specify a branch in a MELPA recipe, I was given guidance from MELPA to _not_ do this, which then points to the default repo branch. At current it's not clear what to do here.

purcell commented:
The reason for the MELPA feedback was that we actively _want_ to build regular MELPA packages off the branch where development is happening, and explicitly-named branches in authors' repositories occasionally get abandoned and go stale when the author changes their development practices.

For curated releases, you add version tags which trigger MELPA Stable versions, and those tags can go on whatever branch. So if you're set on having two separate branches, I'd suggest you should have your default branch be for development (MELPA would build from that), and then a separate release branch to which you merge periodically and then tag.

But really, at this scale it makes little practical sense to have two branches, and you'd have an easier life doing development directly in `main` (with occasional feature branches as you work on stuff) and tagging it periodically.

kickingvegas commented:
@purcell - Thanks for the clarification. While I’ve been using MELPA for many years, it never sunk in until now that all that all the packages there are in a *development* state. Further complicating this is the [guidance](https://github.com/melpa/melpa/blob/5ffb32a540ebf6ebffd1913de1d82019674177fe/README.md?plain=1#L84) to not use MELPA Stable which to me seems odd: I’d think that most users would want to use the stable builds as they would not want to be exposed to untested or worse, unstable behavior from a development build and for that the MELPA maintainers would steer them there.

As most users have been steered to MELPA, in my mind that effectively turns it into a “production” destination, diminishing the value of MELPA stable.

Given the current circumstances, I’m not sanguine about PRs direct to `main` as they would be published directly on MELPA, effectively going to “production”.

As far as your PR goes, I’m good with trying this out in `development` and will try to test it with the guidance you’ve offered. If all goes well it will go out on `main` in the next 1.3.0 release.

purcell commented:
> While I’ve been using MELPA for many years, it never sunk in until now that all the packages there are in a development state.

You've probably also not often encountered unexpected breakage. That's been my experience.

> I’d think that most users would want to use the stable builds as they would not want to be exposed to untested or worse, unstable behavior from a development build and for that the MELPA maintainers would steer them there.

The trouble with MELPA Stable — and why I don't personally encourage it — is that many packages there have dependencies which themselves don't have stable releases, ie. they only exist in regular MELPA. So it's a façade of stability, because those dependencies could change at any time, and also some packages simply don't see regular updates. You _can_ technically set up `package.el` to prefer MELPA Stable releases where they exist, if you like.

> Given the current circumstances, I’m not sanguine about PRs direct to main as they would be published directly on MELPA, effectively going to “production”.

Like all the other packages, which rarely see problems. And not so much "going to production" but "going to a place where people expect to find snapshot versions of this package". /shrug

kickingvegas commented:
@purcell Thanks for the guidance. Merging your PR.

purcell commented:
Great, thanks!

kickingvegas committed:
Redesign Dired link menu and cleanup Makefiles

- Redesign link menu to support the following types of directory links:
  - Absolute Symbolic
  - Relative Symbolic
  - Hard
- Remove "Symlink…" (S) menu item from `casual-dired-tmenu`.
- Add "Link›" (l) menu item to `casual-dired-tmenu`.
- Update screenshots and documentation.
- Cleanup Makefiles to consolidate common variables, rules, and targets.
